### ISSUE TITLE
Multiple checkout intermittent error on self-hosted Linux agents using containers

### ISSUE THREAD

**Issue by mtene** [microsoft/azure-pipelines-agent#3520]
Given the following checkout steps

```
- checkout: main_repo
  path: s

- checkout: self    # this is extension_repo, where the pipeline definition is hosted
  path: s/src/.../subdir
```

our self-hosted agents intermittently throw this error on the second checkout:
```
##[error]Unable to process command '##vso[plugininternal.updaterepositorypath alias=self;]/datadrive/vsts/agent/_work/117/s/src/.../subdir' successfully. Please reference documentation (http://go.microsoft.com/fwlink/?LinkId=817296)
##[error]Repository path '/datadrive/vsts/agent/_work/117/s/src/.../subdir' should be located under agent's work directory '/__w/117'.
Repository is current at '/datadrive/vsts/agent/_work/117/s/extension_repo', move to '/datadrive/vsts/agent/_work/117/s/src/.../subdir'.
##[warning]Unable move and reuse existing repository to required location.
Repository will be located at '/datadrive/vsts/agent/_work/117/s/src/.../subdir'.
```

The full checkout that follows is successful, but the error stops the pipeline from continuing.
If a re-queue of this pipeline catches the same agent, then the problem goes away, since the checkout will be incremental from what was done in the faulty run. Unfortunately, the problem will reappear maybe 1-2 weeks later, probably after the agent undergoes maintenance.

It might be relevant that extension_repo is a git submodule of main_repo, located exactly at the path where the checkout happens. The reason we are doing it this way is because this is a CI pipeline for extension_repo, so we need to checkout the version of extension_repo corresponding to the PR being tested. The code under extension_repo is an optional plugin to main_repo, so it can not be tested in isolation.

**Comment by EzzhevNikita** [comment]
@mtene, thank you for logging this ticket.
I'm investigating the issue now. And I need additional info about the issue.
- What is agent type do you use to repro this issue? Is it Linux self-hosted or Docker container-based self-hosted agent?
- Could you please share YAML of the affected pipeline or a YAML with a minimal reproducible sample to allow me repro this on my side?
- Could you please share the full raw log of affected job (with enabled "Enable system diagnostic")?

**Comment by mtene** [comment]
Hi again @EzzhevNikita I can reproduce the issue with the following pipeline definition (the second job fails):

```
trigger: none

resources:
  repositories:
    - repository: other_repo
      name: <some repo other than self>
      type: git
  containers:
    - container: my_container
      image: <...>
      endpoint: <...>
      options: --cap-add=SYS_PTRACE --security-opt seccomp=unconfined --shm-size=1g

parameters:
- name: agent
  type: string
  default: <agent name>
- name: pool
  type: string
  default: <pool name>

stages:
- stage: Checkout
  jobs:
  # Clean checkout to the default paths
  - job: InitialCheckout
    container: my_container
    pool:
      name: ${{ parameters.pool }}
      demands:
      - Agent.ComputerName -equals ${{ parameters.agent }}
      - Agent.OS -equals Linux
      - docker
    workspace:
      clean: all
    steps:
    - checkout: other_repo
    - checkout: self
    - script: ls -lha $(Build.SourcesDirectory)

  # Incremental checkout to the desired paths
  - job: BuggyCheckout
    dependsOn: InitialCheckout
    continueOnError: true
    container: my_container
    pool: 
      name: ${{ parameters.pool }}
      demands:
      - Agent.ComputerName -equals ${{ parameters.agent }}
      - Agent.OS -equals Linux
      - docker
    steps:
    - checkout: other_repo
      path: s
    - checkout: self
      path: s/subdir
    - script: ls -lha $(Build.SourcesDirectory)

  # Cleanup
  - job: Cleanup
    dependsOn: BuggyCheckout
    container: my_container
    pool: 
      name: ${{ parameters.pool }}
      demands:
      - Agent.ComputerName -equals ${{ parameters.agent }}
      - Agent.OS -equals Linux
      - docker
    workspace:
      clean: all
    steps:
    - script: ls -lha $(Build.SourcesDirectory)
```

I have the logs with diagnostics on, which I can share privately.
This extract might be relevant:
```
##[error]Unable to process command '##vso[plugininternal.updaterepositorypath alias=self;]/datadrive/vsts/agent/_work/31/s/subdir' successfully. Please reference documentation (http://go.microsoft.com/fwlink/?LinkId=817296)
##[error]Repository path '/datadrive/vsts/agent/_work/31/s/subdir' should be located under agent's work directory '/__w/31'.
##[debug]System.ArgumentException: Repository path '/datadrive/vsts/agent/_work/31/s/subdir' should be located under agent's work directory '/__w/31'.
   at Microsoft.VisualStudio.Services.Agent.Worker.Build.BuildDirectoryManager.GetRelativeRepositoryPath(String buildDirectory, String repositoryPath)
   at Microsoft.VisualStudio.Services.Agent.Worker.PluginInternalUpdateRepositoryPathCommand.Execute(IExecutionContext context, Command command)
   at Microsoft.VisualStudio.Services.Agent.Worker.BaseWorkerCommandExtension.ProcessCommand(IExecutionContext context, Command command)
   at Microsoft.VisualStudio.Services.Agent.Worker.WorkerCommandManager.TryProcessCommand(IExecutionContext context, String input)
##[debug]Processed: ##vso[plugininternal.updaterepositorypath alias=self;]/datadrive/vsts/agent/_work/31/s/subdir
##[debug]Repository requires to be placed at '/datadrive/vsts/agent/_work/31/s/subdir', current location is '/datadrive/vsts/agent/_work/31/s/self_repo'
Repository is current at '/datadrive/vsts/agent/_work/31/s/self_repo', move to '/datadrive/vsts/agent/_work/31/s/subdir'.
##[debug]Move existing repository '/datadrive/vsts/agent/_work/31/s/self_repo' to '/datadrive/vsts/agent/_work/31/s/subdir' via staging directory '/datadrive/vsts/agent/_work/_temp/_1'.
##[debug]Catch exception during repository move.
##[debug]System.IO.DirectoryNotFoundException: Directory not found: '/datadrive/vsts/agent/_work/31/s/self_repo'
   at Microsoft.VisualStudio.Services.Agent.Util.ArgUtil.Directory(String directory, String name)
   at Microsoft.VisualStudio.Services.Agent.Util.IOUtil.MoveDirectory(String sourceDir, String targetDir, String stagingDir, CancellationToken token)
   at Agent.Plugins.Repository.CheckoutTask.RunAsync(AgentTaskPluginExecutionContext executionContext, CancellationToken token)
##[warning]Unable move and reuse existing repository to required location.
##[debug]Processed: ##vso[task.logissue type=warning;]Unable move and reuse existing repository to required location.
Repository will be located at '/datadrive/vsts/agent/_work/31/s/subdir'.
Syncing repository: self_repo (Git)
```

**Comment by denis-tikhomirov** [comment]
Resolved in Agent version 2.195.0.
